packages/opencensus-exporter-stackdriver/src/stackdriver-cloudtrace-utils.ts

```diff
@@ -0,0 +1,170 @@
+/**
+ * Copyright 2019, OpenCensus Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as coreTypes from '@opencensus/core';
+import * as types from './types';
+
+const AGENT_LABEL_KEY = 'g.co/agent';
+const AGENT_LABEL_VALUE_STRING = `opencensus-node [${coreTypes.version}]`;
+const AGENT_LABEL_VALUE = createAttributeValue(AGENT_LABEL_VALUE_STRING);
+
+/**
+ * Creates StackDriver Links from OpenCensus Link.
+ * @param links coreTypes.Link[]
+ * @param droppedLinksCount number
+ * @returns types.Links
+ */
+export function createLinks(
+    links: coreTypes.Link[], droppedLinksCount: number): types.Links {
+  return {link: links.map((link) => createLink(link)), droppedLinksCount};
+}
+
+/**
+ * Creates StackDriver Attributes from OpenCensus Attributes.
+ * @param attributes coreTypes.Attributes
+ * @param resourceLabels Record<string, types.AttributeValue>
+ * @param droppedAttributesCount number
+ * @returns types.Attributes
+ */
+export function createAttributes(
+    attributes: coreTypes.Attributes,
+    resourceLabels: Record<string, types.AttributeValue>,
+    droppedAttributesCount: number): types.Attributes {
+  const attributesBuilder =
+      createAttributesBuilder(attributes, droppedAttributesCount);
+  attributesBuilder.attributeMap[AGENT_LABEL_KEY] = AGENT_LABEL_VALUE;
+  attributesBuilder.attributeMap =
+      Object.assign({}, attributesBuilder.attributeMap, resourceLabels);
+  return attributesBuilder;
+}
+
+/**
+ * Creates StackDriver TimeEvents from OpenCensus Annotation and MessageEvent.
+ * @param annotationTimedEvents coreTypes.Annotation[]
+ * @param messageEventTimedEvents coreTypes.MessageEvent[]
+ * @param droppedAnnotationsCount number
+ * @param droppedMessageEventsCount number
+ * @returns types.TimeEvents
+ */
+export function createTimeEvents(
+    annotationTimedEvents: coreTypes.Annotation[],
+    messageEventTimedEvents: coreTypes.MessageEvent[],
+    droppedAnnotationsCount: number,
+    droppedMessageEventsCount: number): types.TimeEvents {
+  const timeEvents: types.TimeEvent[] = [];
+  if (annotationTimedEvents) {
+    annotationTimedEvents.forEach(annotation => {
```

Comment:
Could this be a `map` rather than `forEach` + `push`?

Reply:
Done.